src/Space.ts

```diff
@@ -1,119 +1,154 @@
+  data: { [key: string]: any };
+};
 
-class MemberUpdateEvent extends Event {
-  constructor(public members: SpaceMember[]) {
-    super('memberUpdate', {});
+type SpaceLeaver = {
+  clientId: string;
+  timeoutId: NodeJS.Timeout;
+};
+
+const SPACE_OPTIONS_DEFAULTS = {
+  offlineTimeout: 120_000,
+};
+
+class SpaceMembersUpdateEvent extends Event {
+  constructor(public message?: Types.PresenceMessage) {
+    super('membersUpdate', {});
   }
 }
 
-const createSpaceMemberFromPresenceMember = (m: Types.PresenceMessage): SpaceMember => ({
-  clientId: m.clientId as string,
-  isConnected: true,
-  data: JSON.parse(m.data as string),
-});
-
 class Space extends EventTarget {
-  private members: SpaceMember[] = [];
   private channelName: string;
+  private clientId: string;
   private channel: Types.RealtimeChannelPromise;
+  private members: SpaceMember[];
+  private leavers: SpaceLeaver[];
+  private options: SpaceOptions;
+
+  eventTarget: EventTarget;
 
-  constructor(private name: string, private client: Types.RealtimePromise, private options?: SpaceOptions) {
+  constructor(private name: string, private client: Types.RealtimePromise, options?: SpaceOptions) {
     super();
+    this.options = { ...SPACE_OPTIONS_DEFAULTS, ...options };
+    this.clientId = this.client.auth.clientId;
+    this.members = [];
+    this.leavers = [];
     this.setChannel(this.name);
   }
 
   private setChannel(rootName) {
-    // The channel name prefix here should be unique to avoid conflicts with non-space channels
-    this.channelName = `_ably_space_${rootName}`;
+    this.channelName = `${SPACE_CHANNEL_PREFIX}${rootName}`;
     this.channel = this.client.channels.get(this.channelName);
   }
 
-  private createMember(clientId: string, isConnected: boolean, data: { [key: string]: any }) {
-    this.members.push({ clientId, isConnected, data });
+  private createSpaceMemberFromPresenceMember(message: Types.PresenceMessage): SpaceMember {
+    return {
+      clientId: message.clientId as string,
+      isConnected: message.action !== 'leave',
+      data: JSON.parse(message.data as string),
+    };
   }
 
-  private syncMembers() {
-    this.channel.presence.get().then((presenceMessages) => {
-      this.members = presenceMessages.filter((m) => m.clientId).map(createSpaceMemberFromPresenceMember);
-    });
+  private mapPresenceMembersToSpaceMembers(messages: Types.PresenceMessage[]) {
+    return messages
+      .filter((message) => message.clientId !== this.clientId)
+      .map((message) => this.createSpaceMemberFromPresenceMember(message));
   }
 
-  private subscribeToPresenceEvents() {
-    this.channel.presence.subscribe('enter', (message: Types.PresenceMessage) => {
-      this.updateMemberState(message.clientId, true, JSON.parse(message.data as string));
+  private addLeaver(message: Types.PresenceMessage) {
+    const timeoutCallback = () => {
+      const membersIndex = this.members.findIndex(({ clientId }) => clientId === message.clientId);
```

Comment:
Managing members & leavers in this way, especially with the mutation involved with splicing, seems like it could become more complicated over time.

Can we implement `members` & `leavers` as a Map of `{ clientId: message }`?

Reply:
We can review later, of course, but I don't think it's especially problematic right now (though I'll happily admit I'm wrong here). We base our mutation on the `clientId`, not the index so no desync there. We're on a single thread so these will always be ordered. Did you have a particular scenario in mind that you can imagine being a problem here?

On that thought, I've realized there is a bug here. `findIndex` will return `-1` if it doesn't find the entity and then we get `.slice(-1,1)` ... which is obviously not something we want.